loader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/Scanner.java

```diff
@@ -34,7 +34,11 @@ public ModFileScanData scan() {
         if (loaders != null) {
             loaders.forEach(loader -> {
                 LOGGER.debug(LogMarkers.SCAN, "Scanning {} with language loader {}", fileToScan.getFilePath(), loader.name());
-                loader.getFileVisitor().accept(result);
+                // Since we now scan files concurrently, but language loaders may not have been written with that in mind,
+                // synchronize on the language loader so it only scans one file at a time
+                synchronized (loader) {
```

Comment:
Doesn't this negate most of the benefits?

Reply:
No, the hot code is `Scanner#fileVisitor`, which still runs concurrently.